langchain_benchmarks/rate_limiting.py

```diff
@@ -0,0 +1,107 @@
+"""Implementation of a rate limiter based on a token bucket."""
+import threading
+import time
+from typing import Any, Optional
+
+from langchain.schema.runnable import Runnable, RunnableLambda
+from langchain.schema.runnable.utils import Input, Output
+
+
+class RateLimiter:
+    def __init__(
+        self,
+        *,
+        requests_per_second: float = 1,
+        tokens_per_request: int = 1,
+        check_every_n_seconds: float = 0.1,
+    ) -> None:
+        """A rate limiter based on a token bucket.
+
+        This is a rate limiter which works in a threaded environment.
+
+        It works by filling up a bucket with tokens at a given rate. Each
+        request consumes a given number of tokens. If there are not enough
+        tokens in the bucket, the request is blocked until there are enough
+        tokens.
+
+        Args:
+            requests_per_second: The number of tokens to add per second to the bucket.
+                Must be at least 1.
+            tokens_per_request: number of tokens that a request "costs".
+                This is 1 by default.
+            check_every_n_seconds: check whether the tokens are available
+                every this many seconds. Can be a float to represent
+                fractions of a second.
+        """
+        if requests_per_second < 1:
+            raise ValueError("Rate must be at least 1 request per second")
```

Comment:
For the lazy user who doesn't read the docstring, thoughts on adding a more instructional recommendation here? I naively would assume I could let there be 1/4 requests per second, for instance